### DISCUSSION TITLE
Issue with Edit externally on Linux/Flatpak?

### DISCUSSION THREAD

**Discussion by DaVince** [ppy/osu#29357] [Q&A]
I'd like to modify some assets for a beatmap I'm working on, but when I get to this screen:

![Screenshot_20240808_225310](https://github.com/user-attachments/assets/0e0dd534-bc9b-4312-8c52-4945b59d668b)

The "Open folder" button does absolutely nothing, and I do not know where the directory's actual location is.

I'm running the [latest Flatpak](https://flathub.org/apps/sh.ppy.osu) version. (Btw, it says Unverified but the developer seems correct, is it official?). On a Steam Deck, if that matters. I've given the Flatpak itself permissions to my user files and even the system files, to no avail.

Is this a known issue? Where should I look in order to find the files? And for any developers, would it perhaps be an idea to also display the path and a way to copy the path? Or a way to figure out why it's not launching the file manager?

**Comment by Stoppedpuma** [comment]
Just to make sure, are you using a terminal based file manager such as Ranger or nnn? Flatpak in particular has issues with terminal based file managers. If you're using Dolphin (the default file manager in KDE) then this is probably just an issue with the Flatpak as the AppImage works fine on my end.

> Btw, it says Unverified but the developer seems correct, is it official?

It's pretty much become official as peppy seems to be the one updating it nowadays.

> Or a way to figure out why it's not launching the file manager?

Uploading logs would be helpful!

**Reply by DaVince** [reply]
Thanks for the reply! Yep, I use Dolphin as the file manager. Also, interestingly, clicking on "Open osu! folder" in the settings _does_ open Dolphin.

Here's the logs. [compressed-logs.zip](https://github.com/user-attachments/files/16573766/compressed-logs.zip)
For the sake of searching through the logs, the beatmap I was trying to edit is named "TROUBLE WANDER!".

Tried the AppImage, and sure enough, it works fine there. So this is definitely a Flatpak permissions issue. I'll mess around with the permissions some more and see if I can get it to work.

**Reply by DaVince** [reply]
I noticed that the AppImage was creating a directory in `/tmp`, so I decided to add that to the Flatpak's permissions. Turned out that is the solution: the directory now opens successfully with the click of a button.

And indeed, after revoking the permissions, when I click on "Edit externally", I don't see any new folders pop up at all in /tmp. I guess it's failing silently and not letting the user know.

(For people coming across this: in KDE you can find "Flatpak Permission Settings" from the application launcher. Scroll to osu!, click "Add new" next to the "Filesystem access" header, enter `/tmp` as the path and give it read/write permissions, then Apply.)

**Reply by DaVince** [reply]
You're right in that it breaks out of the sandbox this way. I can live with it, but perhaps a better solution is in order.

**Reply by Stoppedpuma** [reply]
Access to only /tmp/lazer may be a better way of handling this. Can you see if this works on your end? I'll push a PR to the flatpak if so.

**Reply by DaVince** [reply]
Hmm, I tried it, but the folders aren't "mounted" into /tmp/lazer but directly into /tmp. I'm guessing it could work if it did, though.

**Reply by Stoppedpuma** [reply]
@peppy Any reason in particular externally edited beatmaps aren't exported to `/tmp/lazer` instead?

**Reply by peppy** [reply]
No specific reason, we can change maybe.

**Reply by Stoppedpuma** [reply]
A change would be preferable as access to `/tmp` can break program isolation as well as in this case give a Flatpak program access to other programs. I'll open a PR on the Flatpak side of things for access to `/tmp/lazer`.

**Reply by peppy** [reply]
> A change would be preferable as access to `/tmp` can break program isolation as well as in this case give a Flatpak program access to other programs. I'll open a PR on the Flatpak side of things for access to `/tmp/lazer`.

It should be approached the other way around.

**Reply by DaVince** [reply]
So, what happens next? Does an opposite approach mean that osu! would internally pick the most appropriate place for its temp files?

**Reply by Stoppedpuma** [reply]
> So, what happens next?

I've opened a [PR](https://github.com/flathub/sh.ppy.osu/pull/191) already on the Flatpak side of things. I think peppy may have misread my comment as allowing access to `/tmp` instead of `/tmp/lazer` but I may be wrong? 

> Does an opposite approach mean that osu! would internally pick the most appropriate place for its temp files?

Another solution as you mentioned could be just dropping the folder into the games directory but that kind of breaks standardised file locations (meaning putting temp files in tmp instead of somewhere else). It makes more sense to me to keep it in /tmp/lazer.
